### PR TITLE
Fix AOT warnings in StackExchangeRedis

### LINKED ISSUE #1341: Make `OpenTelemetry.Instrumentation.StackExchangeRedis` package AOT safe.
# Issue with OpenTelemetry.Instrumentation.StackExchangeRedis

Related to: https://github.com/open-telemetry/opentelemetry-dotnet/issues/3429
cc: @vitek-karas, @Yun-Ting, @reyang 

Runtime version: `net6.0`, `net7.0`.

**Is this a feature request or a bug?**

* [x] Feature Request
* [ ] Bug

## Additional Context

Here are the warnings I see when adding this library to the `OpenTelemetry.AotCompatibility.TestApp.csproj` app:

```
C:\git\opentelemetry-dotnet-contrib\src\Shared\ActivityInstrumentationHelper.cs(36): Trim analysis warning IL2026: OpenTelemetry.Instrumentation.ActivityInstrumentationHelper.CreateActivitySourceSetter(): Using member 'System.Linq.Expressions.Expression.Property(Expression,String)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Creating Expressions requires unreferenced code because the members being referenced by the Expression may be trimmed. [C:\git\opentelemetry-dotnet-contrib\test\OpenTelemetry.AotCompatibility.TestApp\OpenTelemetry.AotCompatibility.TestApp.csproj]
C:\git\opentelemetry-dotnet-contrib\src\Shared\ActivityInstrumentationHelper.cs(44): Trim analysis warning IL2026: OpenTelemetry.Instrumentation.ActivityInstrumentationHelper.CreateActivityKindSetter(): Using member 'System.Linq.Expressions.Expression.Property(Expression,String)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Creating Expressions requires unreferenced code because the members being referenced by the Expression may be trimmed. [C:\git\opentelemetry-dotnet-contrib\test\OpenTelemetry.AotCompatibility.TestApp\OpenTelemetry.AotCompatibility.TestApp.csproj]
C:\git\opentelemetry-dotnet-contrib\src\Shared\MultiTypePropertyFetcher.cs(64): Trim analysis warning IL2075: OpenTelemetry.Instrumentation.MultiTypePropertyFetcher`1.Fetch(Object): 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicProperties', 'DynamicallyAccessedMemberTypes.NonPublicProperties' in call to 'System.Reflection.TypeInfo.DeclaredProperties.get'. The return value of method 'System.Object.GetType()' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [C:\git\opentelemetry-dotnet-contrib\test\OpenTelemetry.AotCompatibility.TestApp\OpenTelemetry.AotCompatibility.TestApp.csproj]
C:\git\opentelemetry-dotnet-contrib\src\Shared\MultiTypePropertyFetcher.cs(67): Trim analysis warning IL2075: OpenTelemetry.Instrumentation.MultiTypePropertyFetcher`1.Fetch(Object): 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicProperties', 'DynamicallyAccessedMemberTypes.NonPublicProperties' in call to 'System.Type.GetProperty(String,BindingFlags)'. The return value of method 'System.Object.GetType()' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [C:\git\opentelemetry-dotnet-contrib\test\OpenTelemetry.AotCompatibility.TestApp\OpenTelemetry.AotCompatibility.TestApp.csproj]
C:\git\opentelemetry-dotnet-contrib\src\Shared\PropertyFetcher.cs(79): Trim analysis warning IL2075: OpenTelemetry.Instrumentation.PropertyFetcher`1.TryFetch(Object,!0&): 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicProperties', 'DynamicallyAccessedMemberTypes.NonPublicProperties' in call to 'System.Reflection.TypeInfo.DeclaredProperties.get'. The return value of method 'System.Object.GetType()' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [C:\git\opentelemetry-dotnet-contrib\test\OpenTelemetry.AotCompatibility.TestApp\OpenTelemetry.AotCompatibility.TestApp.csproj]
C:\git\opentelemetry-dotnet-contrib\src\Shared\PropertyFetcher.cs(82): Trim analysis warning IL2075: OpenTelemetry.Instrumentation.PropertyFetcher`1.TryFetch(Object,!0&): 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicProperties' in call to 'System.Type.GetProperty(String)'. The return value of method 'System.Object.GetType()' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [C:\git\opentelemetry-dotnet-contrib\test\OpenTelemetry.AotCompatibility.TestApp\OpenTelemetry.AotCompatibility.TestApp.csproj]
C:\git\opentelemetry-dotnet-contrib\src\OpenTelemetry.Instrumentation.StackExchangeRedis\Implementation\RedisProfilerEntryToActivityConverter.cs(191): Trim analysis warning IL2070: OpenTelemetry.Instrumentation.StackExchangeRedis.Implementation.RedisProfilerEntryToActivityConverter.CreateFieldGetter<TField>(Type,String,BindingFlags): 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicFields', 'DynamicallyAccessedMemberTypes.NonPublicFields' in call to 'System.Type.GetField(String,BindingFlags)'. The parameter 'classType' of method 'OpenTelemetry.Instrumentation.StackExchangeRedis.Implementation.RedisProfilerEntryToActivityConverter.CreateFieldGetter<TField>(Type,String,BindingFlags)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [C:\git\opentelemetry-dotnet-contrib\test\OpenTelemetry.AotCompatibility.TestApp\OpenTelemetry.AotCompatibility.TestApp.csproj]
C:\git\opentelemetry-dotnet-contrib\src\OpenTelemetry.Instrumentation.StackExchangeRedis\Implementation\RedisProfilerEntryToActivityConverter.cs(195): AOT analysis warning IL3050: OpenTelemetry.Instrumentation.StackExchangeRedis.Implementation.RedisProfilerEntryToActivityConverter.CreateFieldGetter<TField>(Type,String,BindingFlags): Using member 'System.Reflection.Emit.DynamicMethod.DynamicMethod(String,Type,Type[],Boolean)' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. Creating a DynamicMethod requires dynamic code. [C:\git\opentelemetry-dotnet-contrib\test\OpenTelemetry.AotCompatibility.TestApp\OpenTelemetry.AotCompatibility.TestApp.csproj]
C:\git\opentelemetry-dotnet-contrib\src\Shared\MultiTypePropertyFetcher.cs(99): AOT analysis warning IL3050: OpenTelemetry.Instrumentation.MultiTypePropertyFetcher`1.PropertyFetch.FetcherForProperty(PropertyInfo): Using member 'System.Type.MakeGenericType(Type[])' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. The native code for this instantiation might not be available at runtime. [C:\git\opentelemetry-dotnet-contrib\test\OpenTelemetry.AotCompatibility.TestApp\OpenTelemetry.AotCompatibility.TestApp.csproj]
C:\git\opentelemetry-dotnet-contrib\src\Shared\PropertyFetcher.cs(107): AOT analysis warning IL3050: OpenTelemetry.Instrumentation.PropertyFetcher`1.PropertyFetch.FetcherForProperty(PropertyInfo): Using member 'System.Type.MakeGenericType(Type[])' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. The native code for this instantiation might not be available at runtime. [C:\git\opentelemetry-dotnet-contrib\test\OpenTelemetry.AotCompatibility.TestApp\OpenTelemetry.AotCompatibility.TestApp.csproj]
C:\git\opentelemetry-dotnet-contrib\src\OpenTelemetry.Instrumentation.StackExchangeRedis\Implementation\RedisProfilerEntryToActivityConverter.cs(31): Trim analysis warning IL2026: OpenTelemetry.Instrumentation.StackExchangeRedis.Implementation.RedisProfilerEntryToActivityConverter.<>c.<.cctor>b__4_0(): Using member 'System.Reflection.Assembly.GetType(String)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Types might be removed. [C:\git\opentelemetry-dotnet-contrib\test\OpenTelemetry.AotCompatibility.TestApp\OpenTelemetry.AotCompatibility.TestApp.csproj]
C:\git\opentelemetry-dotnet-contrib\src\OpenTelemetry.Instrumentation.StackExchangeRedis\Implementation\RedisProfilerEntryToActivityConverter.cs(32): Trim analysis warning IL2026: OpenTelemetry.Instrumentation.StackExchangeRedis.Implementation.RedisProfilerEntryToActivityConverter.<>c.<.cctor>b__4_0(): Using member 'System.Reflection.Assembly.GetType(String)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Types might be removed. [C:\git\opentelemetry-dotnet-contrib\test\OpenTelemetry.AotCompatibility.TestApp\OpenTelemetry.AotCompatibility.TestApp.csproj]
/_/src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/ArrayOperations.cs(23): AOT analysis warning IL3050: System.Linq.Expressions.Interpreter.NewArrayInitInstruction.Run(InterpretedFrame): Using member 'System.Array.CreateInstance(Type,Int32)' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. The code for an array of the specified type might not be available. [C:\git\opentelemetry-dotnet-contrib\test\OpenTelemetry.AotCompatibility.TestApp\OpenTelemetry.AotCompatibility.TestApp.csproj]
/_/src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/ArrayOperations.cs(52): AOT analysis warning IL3050: System.Linq.Expressions.Interpreter.NewArrayInstruction.Run(InterpretedFrame): Using member 'System.Array.CreateInstance(Type,Int32)' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. The code for an array of the specified type might not be available. [C:\git\opentelemetry-dotnet-contrib\test\OpenTelemetry.AotCompatibility.TestApp\OpenTelemetry.AotCompatibility.TestApp.csproj]
/_/src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/ArrayOperations.cs(87): AOT analysis warning IL3050: System.Linq.Expressions.Interpreter.NewArrayBoundsInstruction.Run(InterpretedFrame): Using member 'System.Array.CreateInstance(Type,Int32[])' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. The code for an array of the specified type might not be available. [C:\git\opentelemetry-dotnet-contrib\test\OpenTelemetry.AotCompatibility.TestApp\OpenTelemetry.AotCompatibility.TestApp.csproj]
/_/src/libraries/System.Linq.Expressions/src/System/Dynamic/Utils/TypeUtils.cs(28): AOT analysis warning IL3050: System.Dynamic.Utils.TypeUtils.GetNullableType(Type): Using member 'System.Type.MakeGenericType(Type[])' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. The native code for this instantiation might not be available at runtime. [C:\git\opentelemetry-dotnet-contrib\test\OpenTelemetry.AotCompatibility.TestApp\OpenTelemetry.AotCompatibility.TestApp.csproj]
/_/src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/NewArrayExpression.cs(228): AOT analysis warning IL3050: System.Linq.Expressions.Expression.NewArrayBounds(Type,IEnumerable`1<Expression>): Using member 'System.Type.MakeArrayType()' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. The code for an array of the specified type might not be available. [C:\git\opentelemetry-dotnet-contrib\test\OpenTelemetry.AotCompatibility.TestApp\OpenTelemetry.AotCompatibility.TestApp.csproj]
/_/src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/NewArrayExpression.cs(232): AOT analysis warning IL3050: System.Linq.Expressions.Expression.NewArrayBounds(Type,IEnumerable`1<Expression>): Using member 'System.Type.MakeArrayType(Int32)' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. The code for an array of the specified type might not be available. [C:\git\opentelemetry-dotnet-contrib\test\OpenTelemetry.AotCompatibility.TestApp\OpenTelemetry.AotCompatibility.TestApp.csproj]
/_/src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/NewArrayExpression.cs(172): AOT analysis warning IL3050: System.Linq.Expressions.Expression.NewArrayInit(Type,IEnumerable`1<Expression>): Using member 'System.Type.MakeArrayType()' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. The code for an array of the specified type might not be available. [C:\git\opentelemetry-dotnet-contrib\test\OpenTelemetry.AotCompatibility.TestApp\OpenTelemetry.AotCompatibility.TestApp.csproj]
/_/src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/BinaryExpression.cs(2239): AOT analysis warning IL3050: System.Linq.Expressions.Expression.GetResultTypeOfShift(Type,Type): Using member 'System.Type.MakeGenericType(Type[])' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. The native code for this instantiation might not be available at runtime. [C:\git\opentelemetry-dotnet-contrib\test\OpenTelemetry.AotCompatibility.TestApp\OpenTelemetry.AotCompatibility.TestApp.csproj]
/_/src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/MethodCallExpression.cs(1379): AOT analysis warning IL3050: System.Linq.Expressions.Expression.ApplyTypeArgs(MethodInfo,Type[]): Using member 'System.Reflection.MethodInfo.MakeGenericMethod(Type[])' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. The native code for this instantiation might not be available at runtime. [C:\git\opentelemetry-dotnet-contrib\test\OpenTelemetry.AotCompatibility.TestApp\OpenTelemetry.AotCompatibility.TestApp.csproj]
```

### PR DESCRIPTION
**Pull request by eerhardt**
- Add net6.0 target, so we can annotate the code correctly and use new APIs.
- Copy PropertyFetcher from open-telemetry/opentelemetry-dotnet which was made AOT compatible. Made a new shared version that other libraries can opt into using over time, so this change has no affect on other libraries.
- Add StackExchangeRedis to the AotCompatibility.TestApp.
- Address warnings to make the code work under AOT/trimmed.

Fix #1341

cc @Yun-Ting @utpilla @vitek-karas

### CODE REVIEW

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryToActivityConverter.cs [hunk 1]
```diff
@@ -20,16 +20,21 @@
...
 {
+    private const string ScriptEvalMessageTypeName = "StackExchange.Redis.RedisDatabase+ScriptEvalMessage";
+
     private static readonly Lazy<Func<object, (string?, string?)>> MessageDataGetter = new(() =>
     {
-        var redisAssembly = typeof(IProfiledCommand).Assembly;
-        Type profiledCommandType = redisAssembly.GetType("StackExchange.Redis.Profiling.ProfiledCommand");
-        Type scriptMessageType = redisAssembly.GetType("StackExchange.Redis.RedisDatabase+ScriptEvalMessage");
+        Type profiledCommandType = Type.GetType("StackExchange.Redis.Profiling.ProfiledCommand, StackExchange.Redis", throwOnError: true)!;
+        Type scriptMessageType = Type.GetType(ScriptEvalMessageTypeName + ", StackExchange.Redis", throwOnError: true)!;
```
Comment (utpilla) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryToActivityConverter.cs:37]: Remove ` + `?

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/OpenTelemetry.Instrumentation.StackExchangeRedis.csproj [hunk 2]
```diff
@@ -1,17 +1,18 @@
...
-    <IncludeSharedInstrumentationSource>true</IncludeSharedInstrumentationSource>
     <IncludeSharedExceptionExtensionsSource>true</IncludeSharedExceptionExtensionsSource>
     <MinVerTagPrefix>Instrumentation.StackExchangeRedis-</MinVerTagPrefix>
     <ImplicitUsings>true</ImplicitUsings>
   </PropertyGroup>
 
   <ItemGroup>
     <Compile Include="$(RepoRoot)\src\Shared\Guard.cs" Link="Includes\Guard.cs" />
+    <Compile Include="$(RepoRoot)\src\Shared\PropertyFetcher.AOT.cs" Link="Includes\PropertyFetcher.AOT.cs" />
+    <Compile Include="$(RepoRoot)\src\Shared\SemanticConventions.cs" Link="Includes\SemanticConventions.cs" />
```
Comment (utpilla) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/OpenTelemetry.Instrumentation.StackExchangeRedis.csproj:15]: Why is this added?

Comment (eerhardt) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/OpenTelemetry.Instrumentation.StackExchangeRedis.csproj:15]: SemanticConventions is used by this assembly. I needed to explicitly include it because I removed `IncludeSharedInstrumentationSource` at the top. `IncludeSharedInstrumentationSource` brings in a bunch of shared source, but the only ones used by this project are ProeprtyFetcher and SemanticConventions.

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryToActivityConverter.cs (hunk 1 above)
Comment (eerhardt) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/Implementation/RedisProfilerEntryToActivityConverter.cs:37]: The AOT/trimming tools don't recognize patterns like 

```C#
var redisAssembly = typeof(IProfiledCommand).Assembly;
redisAssembly.GetType("StackExchange.Redis.Profiling.ProfiledCommand");
```

The don't know which Type you are referring to when you do it this way. But if you instead say:

```C#
Type.GetType("StackExchange.Redis.Profiling.ProfiledCommand, StackExchange.Redis");
```

They do recognize this pattern and know exactly which type you are talking about.

The `+` is added above because I refactored out the `StackExchange.Redis.RedisDatabase+ScriptEvalMessage` name into a constant string so I could use it in 2 places:

1. In the Type.GetType call
2. In the `[DynamicDependency("CommandAndKey", ScriptEvalMessageTypeName, "StackExchange.Redis")]` line.

File: src/Shared/PropertyFetcher.AOT.cs [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+#nullable disable
+
+// NOTE: This version of PropertyFetcher is AOT-compatible.
+// Usages of the non-AOT-compatible version can be moved over to this one when they need to support AOT/trimming.
+
```
Comment (eerhardt) [inline src/Shared/PropertyFetcher.AOT.cs:21]:
```suggestion
// Copied from https://github.com/open-telemetry/opentelemetry-dotnet/blob/86a6ba0b7f7ed1f5e84e5a6610e640989cd3ae9f/src/Shared/DiagnosticSourceInstrumentation/PropertyFetcher.cs#L30

```

File: test/OpenTelemetry.AotCompatibility.TestApp/OpenTelemetry.AotCompatibility.TestApp.csproj [hunk 4]
```diff
@@ -11,6 +11,7 @@
   <ItemGroup>
     <TrimmerRootAssembly Include="OpenTelemetry.Instrumentation.Runtime" />
     <TrimmerRootAssembly Include="OpenTelemetry.Instrumentation.EventCounters" />
+    <TrimmerRootAssembly Include="OpenTelemetry.Instrumentation.StackExchangeRedis" />
```
Comment (utpilla) [inline test/OpenTelemetry.AotCompatibility.TestApp/OpenTelemetry.AotCompatibility.TestApp.csproj:14]: Could you also update the [ci.yml](https://github.com/CodeBlanch/opentelemetry-dotnet-contrib/blob/main/.github/workflows/ci.yml#L271-L279) file to add a contains check for the key `redis` as well?

Similar to this change https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1406/files#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR289

File: src/Shared/PropertyFetcher.AOT.cs (hunk 3 above)
Comment (eerhardt) [reply src/Shared/PropertyFetcher.AOT.cs:21]: I've opened https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/1416 to remove the old version and update the other references to this version.

**Review by utpilla — APPROVED** [review]
Please add a CHANGELOG entry.

**Review by vitek-karas — APPROVED** [review]

**Comment by eerhardt** [comment]
I've addressed all existing feedback. I believe this PR is ready for review.

File: src/Shared/PropertyFetcher.AOT.cs [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+                var declaringType = propertyInfo.DeclaringType;
+                if (declaringType!.IsValueType)
+                {
+                    throw new NotSupportedException(
+                        $"Type: {declaringType.FullName} is a value type. PropertyFetcher can only operate on reference payload types.");
+                }
+
+                if (declaringType == typeof(object))
+                {
+                    // TODO: REMOVE this if branch when .NET 7 is out of support.
```
Comment (Yun-Ting) [inline src/Shared/PropertyFetcher.AOT.cs:136]: net7.0 is not one of the target for the project currently:
https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1415/files#:~:text=%3CTargetFrameworks%3Enet6.0%3Bnetstandard2.0%3Bnet462%3C/TargetFrameworks%3E
Do we need this here?

**Comment by Yun-Ting** [comment]
LGTM. Left a nit comment. Thanks!

**Review by Yun-Ting — APPROVED** [review]

File: src/Shared/PropertyFetcher.AOT.cs (hunk 5 above)
Comment (eerhardt) [reply src/Shared/PropertyFetcher.AOT.cs:136]: I would like this to be a straight copy from opentelemetry-dotnet. Then we can keep the 2 in sync over time.

Also, the reason this code is here isn't in case one of the libraries targets net7.0, its for when the application targets net7.0, which we can't control here.

File: test/OpenTelemetry.AotCompatibility.TestApp/OpenTelemetry.AotCompatibility.TestApp.csproj (hunk 4 above)
Comment (CodeBlanch) [reply test/OpenTelemetry.AotCompatibility.TestApp/OpenTelemetry.AotCompatibility.TestApp.csproj:14]: Thanks @utpilla! I just opened #1422 to try and smooth this out a bit.
